{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import qiime2\n",
    "from tempfile import mkdtemp\n",
    "from qiime2.plugins import demux, deblur, quality_filter, \\\n",
    "                           metadata, feature_table, alignment, \\\n",
    "                           phylogeny, diversity, emperor, feature_classifier, \\\n",
    "                           taxa, composition\n",
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np\n",
    "import random\n",
    "import logging\n",
    "from datetime import datetime\n",
    "from Bio import Entrez\n",
    "from pprint import pprint\n",
    "from sklearn.utils import shuffle\n",
    "import xml.etree.ElementTree as ET\n",
    "import os\n",
    "from tqdm.notebook import tqdm\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Studies\n",
    "* exercise_frequency\n",
    "* flossing_frequency\n",
    "* vitamin_d_supplement_frequency\n",
    "* weight_change\n",
    "* fruit_frequency"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pipeline cleaning"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Puliscel a working directory (result_extraction), elimina tutti i file e le cartelle fatta eccezione per il logfile\n",
    "# il quale viene solo svuotato\n",
    "def clean_workspace():\n",
    "    \n",
    "    # ci spostiamo nella working directory\n",
    "    starting_path = os.getcwd()\n",
    "    path = os.path.join(os.getcwd(), f'result_extraction')\n",
    "    os.chdir(path)\n",
    "    \n",
    "    # prendiamo tutti i file contenuti nella working directory e eliminiamo a meno che non sia il log file\n",
    "    files = os.listdir()\n",
    "    for file in files:\n",
    "        if file != 'sra_querying.log':\n",
    "            delete_command = f'rm -r {file}'\n",
    "            os.system(delete_command)\n",
    "            \n",
    "    # puliamo il log file\n",
    "    delete_old_log = 'cat > sra_querying.log'\n",
    "    os.system(delete_old_log)  \n",
    "    \n",
    "    # torniamo nella starting directory ./Human microbiome\n",
    "    os.chdir(starting_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### NCBI utilities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Funzioni per semplificare l'interrogazioen di NCBI\n",
    "Entrez.email = \"giacomo.villa.mi@gmail.com\"\n",
    "\n",
    "def good_print(text):\n",
    "    print(pprint(text))\n",
    "\n",
    "def esearch(db, query, num_max = 20):\n",
    "    handle = Entrez.esearch(db = db, term = query, retmax = num_max)\n",
    "    record = Entrez.read(handle, validate = True)\n",
    "    return record\n",
    "\n",
    "def esummary(db, id_val):\n",
    "    handle = Entrez.esummary(db = db, id = id_val)\n",
    "    record = Entrez.read(handle, validate = True)\n",
    "    return record"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Final Sample generation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prende in input il dataset completo, data la query, \n",
    "def write_age_mean(total, typology, experiment):\n",
    "    \n",
    "    man = total.query('sex == \"male\"')\n",
    "    woman = total.query('sex == \"female\"')\n",
    "    \n",
    "    mean_age_total = round(np.mean(total['age_years']), 4)\n",
    "    mean_age_man = round(np.mean(man['age_years']), 4)\n",
    "    mean_age_woman = round(np.mean(woman['age_years']), 4)\n",
    "    not_valid_sex = total.query(\"sex != 'female' and sex != 'male'\")\n",
    "    \n",
    "    if typology != 'sample':\n",
    "        logging.info(f'--------------------{typology.upper()}--------------------')\n",
    "    logging.info(f'Total number of {typology} people: {len(total)}')\n",
    "    logging.info(f'Total number of {typology} man: {len(man)}')\n",
    "    logging.info(f'Total number of {typology} woman: {len(woman)}')\n",
    "    logging.info(f'Total number of non valid sex {len(not_valid_sex)}')\n",
    "    \n",
    "    logging.info(f'Mean Age for total: {mean_age_total}')\n",
    "    logging.info(f'Mean Age for man: {mean_age_man}')\n",
    "    logging.info(f'Mean Age for woman: {mean_age_woman}')\n",
    "    logging.info(f'\\n')\n",
    "    \n",
    "    print(f'--------------------{typology.upper()}--------------------')\n",
    "    print(f'Total number of {typology} people: {len(total)}')\n",
    "    print(f'Total number of {typology} man: {len(man)}')\n",
    "    print(f'Total number of {typology} woman: {len(woman)}')\n",
    "    print(f'Total number of non valid sex {len(not_valid_sex)}')\n",
    "    \n",
    "    print(f'Mean Age for total: {mean_age_total}')\n",
    "    print(f'Mean Age for man: {mean_age_man}')\n",
    "    print(f'Mean Age for woman: {mean_age_woman}')\n",
    "    \n",
    "    starting_path = os.getcwd()\n",
    "    path = os.path.join(os.getcwd(), f'result_extraction/')\n",
    "    os.chdir(path)\n",
    "    directories = os.listdir()\n",
    "    if f'{experiment}' not in directories:\n",
    "        os.mkdir(f'{experiment}')\n",
    "        \n",
    "    os.chdir(f'{experiment}')\n",
    "    if typology != 'sample':\n",
    "        total.to_csv(f\"./dataset_query_result_{typology}.csv\", index=False, encoding='utf-8')\n",
    "    \n",
    "    os.chdir(starting_path)\n",
    "\n",
    "# Prende in input il dataset da cui campionare e gli index degli element già presi, finché non campiona qualcosa di \n",
    "# nuovo non termina\n",
    "def generate_single_sample(already_taken, from_df):\n",
    "    element = random.randint(0, len(from_df) - 1)\n",
    "    while (element in already_taken):\n",
    "        element = random.randint(0, len(from_df) - 1)\n",
    "    return element\n",
    "\n",
    "# Dato il dataset di partenza contenente solo elementi effettivamente presenti su NCBI\n",
    "# genera un nuovo dataset composto da n_samples elementi\n",
    "def get_final_sample(started_dataset, n_samples):\n",
    "    taken = set()\n",
    "    final_sample = pd.DataFrame(columns=started_dataset.columns)\n",
    "    for i in tqdm(range(n_samples), desc='Sampling data'):\n",
    "        new_sample = generate_single_sample(taken, started_dataset)\n",
    "        taken.add(new_sample)\n",
    "        final_sample = final_sample.append(started_dataset.iloc[new_sample], ignore_index=True)\n",
    "    return final_sample\n",
    "\n",
    "def write_not_valid_ids(not_valid_names):\n",
    "    files = os.listdir(f'./result_extraction')\n",
    "    \n",
    "    if 'not_valid_sample_names.csv' in files:\n",
    "        all_not_valid_names = pd.read_csv(\"./result_extraction/not_valid_sample_names.csv\", header=0, dtype=str)\n",
    "        elements = [all_not_valid_names, not_valid_names]\n",
    "        final_not_valid = pd.concat(elements, ignore_index=False, sort=False)\n",
    "        final_not_valid.to_csv(f\"./result_extraction/not_valid_sample_names.csv\", index=False)\n",
    "    else:\n",
    "        not_valid_names.to_csv(f\"./result_extraction/not_valid_sample_names.csv\", index=False)\n",
    "    \n",
    "# Dato il dataset di partenza con tutti gli elementi che rispettano la query posta al gut, tiene in considraz\n",
    "def sampling_data(start_dataset, typology, n_samples, experiment):\n",
    "    write_age_mean(start_dataset, typology, experiment)\n",
    "    valid_id = list()\n",
    "    not_valid_id = list()\n",
    "    for try_id in tqdm(start_dataset['sample_name'], desc='NCBI ids validation'):\n",
    "        handleSce = esearch('biosample', try_id)\n",
    "        if len(handleSce['IdList']) != 0:\n",
    "            valid_id.append(try_id)\n",
    "        else:\n",
    "            not_valid_id.append(str(try_id))\n",
    "            print(try_id)\n",
    "            \n",
    "    print(f'Total rows: {len(start_dataset)}')\n",
    "    print(f'Valid rows: {len(valid_id)}')\n",
    "    \n",
    "    not_valid_names = pd.DataFrame(data={\"not_valid_sample_name\": not_valid_id, \"typology\": [typology]*len(not_valid_id)})\n",
    "    print(len(not_valid_names))\n",
    "    write_not_valid_ids(not_valid_names)\n",
    "    \n",
    "    logging.info(f'Total rows: {len(start_dataset)}')\n",
    "    logging.info(f'Valid rows: {len(valid_id)}')\n",
    "            \n",
    "    valid_start_dataset = pd.DataFrame(columns=start_dataset.columns)\n",
    "    index = 0\n",
    "    for _, row in start_dataset.iterrows():\n",
    "        if row['sample_name'] in valid_id:\n",
    "            valid_start_dataset.loc[index] = row\n",
    "            index += 1\n",
    "            \n",
    "    valid_start_dataset = shuffle(valid_start_dataset)\n",
    "    final_sample = get_final_sample(valid_start_dataset, n_samples)\n",
    "    write_sample_info(final_sample, typology, experiment)\n",
    "\n",
    "# Funzione richiamata sul sample finale, traduce in csv il campione finale creando un csv di due colonne: sample_name,\n",
    "# typology (e.g. healthy/not_healthy), richiama write_age_mean per scrivere sul log file l'età media dei soggetti\n",
    "def write_sample_info(sample, typology, experiment):\n",
    "    man = sample.query(\"sex == 'male'\")\n",
    "    woman = sample.query(\"sex == 'female'\")\n",
    "    write_age_mean(sample, 'sample', experiment)\n",
    "    \n",
    "    sample = sample[['sample_name']]\n",
    "    sample['typology'] = [typology]*len(sample)\n",
    "    \n",
    "    files = os.listdir(f'./result_extraction/{experiment}')\n",
    "    \n",
    "    if f'final_sample_{experiment}.csv' in files:\n",
    "        final_sample = pd.read_csv(f\"./result_extraction/{experiment}/final_sample_{experiment}.csv\", header=0, dtype=str)\n",
    "        final_sample = final_sample[['sample_name', 'typology']]\n",
    "        elements = [final_sample, sample]\n",
    "        final_sample = pd.concat(elements, ignore_index=False, sort=False)\n",
    "        final_sample.to_csv(f\"./result_extraction/{experiment}/final_sample_{experiment}.csv\", index=False)\n",
    "    else:\n",
    "        sample.to_csv(f\"./result_extraction/{experiment}/final_sample_{experiment}.csv\", index=False)\n",
    "        \n",
    "    close_dashes = '-'*len(typology.upper())\n",
    "    logging.info(f'--------------------{close_dashes}--------------------')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SRA operation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Funzione controller, prende in input il nome dell'esperimento (eg. healthy vs not_healthy) e le tipologie di campione\n",
    "# (e.g. healthy e not_healty), chiama la funzione che interroga SRA di NCBI, in seguito richiama la funzione per \n",
    "# concatenare i file fasta e infine la funzione per prendere le sequenze che ricorrono più spesso. Infine salva su un\n",
    "# file csv i fasta delle sequenze più popolose aggiungendo il campo che si rifà all'id di Biosample.\n",
    "# Il file csv delle frequenze più popole è l'input per blast\n",
    "def sra_querying(experiment, types):\n",
    "    \n",
    "    # Legge il csv contente i 30 campioni di una tipologia e i 30 campioni dell'altra tipologia dato l'esperimento\n",
    "    # (e.g. esperimento: healthy vs not_healthy estrare il csv che contiene i 30 sample_name degli healthy e i \n",
    "    # 30 sample_name dei not_healthy)\n",
    "    final_sample = pd.read_csv(f\"./result_extraction/{experiment}/final_sample_{experiment}.csv\", header=0, dtype=str)\n",
    "    \n",
    "    # crea una lista che conterrà, dati i record, il corrispettivo id di Biosample\n",
    "    bio_sample_id = list()\n",
    "    \n",
    "    # per ogni riga del csv dei 30 campioni di una tipologia e i 30 dell'altra tipologia dato l'esperimento\n",
    "    for index, row in final_sample.iterrows():\n",
    "        print(f'File number: {index+1}')\n",
    "        \n",
    "        # Gestione del problema sulla lettura di un sample_name con la concatenazione della stringa '001'\n",
    "        # in generale estrae l'input per la funzione che farà la query su SRA\n",
    "        record_id = str(row[0])[0:15]\n",
    "        record_typology = row[1]\n",
    "        \n",
    "        # data la singola interrogazione, aggiunge alla lista degli id di bio_sample l'id.\n",
    "        bio_sample_id.append(get_sequences(record_id, record_typology, experiment))\n",
    "        \n",
    "    # una volta scaricate tutti i file fasta data l'esperimento, per ogni tipologia (e.g healthy/not_healthy) \n",
    "    # crea un unico file con tutte le sequenze e poi prende, da questo file, solo quelle più popolose\n",
    "    for typology in types:\n",
    "        concatenate_fast_file(typology, 'fasta', experiment)\n",
    "        get_top_sequences(typology, experiment)\n",
    "        \n",
    "    # crea una nuova colonna dove, per ogni sample_name, vi sarà l'id di biosample associato e salva il nuovo csv\n",
    "    final_sample['bio_sample_id'] = bio_sample_id\n",
    "    final_sample.to_csv(f\"./result_extraction/{experiment}/final_sample_{experiment}.csv\", index=False)\n",
    "    \n",
    "# Funzione che interroga SRA, dato il sample_name. Richiede anche la tipologia del campione (e.g. healthy o not_healthy)\n",
    "# e il nome dell'esperimento (e.g. healthy_vs_not_healthy), per andare a salvare correttamente nelle cartelle facenti\n",
    "# riferimento all'esperimento\n",
    "def get_sequences(sample_name, typology, experiment):\n",
    "    \n",
    "    # in funzione della tipologia del campione (e.g. healthy o not_healthy) e dell'esperimento (e.g. healthy_vs_not_healthy)\n",
    "    # definisce il path corretto dove andare a salvare il risultato\n",
    "    path = f'\"./result_extraction/{experiment}/SRA_{typology}\" '\n",
    "    command1 = f'fastq-dump --fasta --readids --outdir {path}'\n",
    "    command2 = f'fastq-dump --readids --outdir {path}'\n",
    "        \n",
    "    # Query su SRA e print utili\n",
    "    print(f'Sample id: {sample_name}')\n",
    "    handleSce = esearch('biosample', sample_name)\n",
    "    biosampleId = handleSce['IdList'][0]\n",
    "    print(f'Biosample ID {biosampleId}')\n",
    "    print(f'Typology: {typology}')\n",
    "    handleSra = Entrez.efetch(db='biosample', id=biosampleId, retmode='xml')\n",
    "    root = ET.fromstring(handleSra.read())\n",
    "    identifier = root.findall('.//BioSample//Ids//Id')\n",
    "    for i in identifier:\n",
    "        if i.attrib['db'] == 'SRA':\n",
    "            sraId = i.text\n",
    "    print(f'SRA ID: {sraId}')\n",
    "    handleSra = Entrez.esearch(db='sra', term=sraId)\n",
    "    resultsSra = Entrez.read(handleSra)['IdList']\n",
    "    print(resultsSra)\n",
    "    for s in resultsSra:\n",
    "        handlesngSraId = Entrez.efetch(db='sra', id=s, retmode='xml')\n",
    "        root = ET.fromstring(handlesngSraId.read())\n",
    "        identifier = root.find('.//EXPERIMENT_PACKAGE//RUN_SET//RUN')\n",
    "        runId = identifier.attrib['accession']\n",
    "        os.system(command1 + runId)\n",
    "        os.system(command2 + runId) \n",
    "    print()\n",
    "    return biosampleId\n",
    "        \n",
    "\n",
    "# Dato il risultato delle query su SRA, concatena i file fasta facenti riferimento a una certa tipologia di record\n",
    "# (e.g. healthy o not_healthy) dato un certo esperimento (e.g. healthy_vs_not_healthy)\n",
    "def concatenate_fast_file(typology, file_format, experiment):\n",
    "    \n",
    "    # Prende tutti i file data la tipologia del record (e.g. healthy o not_healthy) contenuti nella cartella dove,\n",
    "    # dato l'esperimento (e.g. healthy_vs_not_healthy), la query su SRA ha riposto i risultati\n",
    "    files = os.listdir(f'./result_extraction/{experiment}/SRA_{typology}')\n",
    "    \n",
    "    # concatenazione file fasta\n",
    "    compact_files = list()\n",
    "    for file in files:\n",
    "        if file_format in file:\n",
    "            f = open(f'./result_extraction/{experiment}/SRA_{typology}/{file}', \"r\")\n",
    "            compact_files.append(f.read())\n",
    "            f.close()\n",
    "    f = open(f'./result_extraction/{experiment}/SRA_{typology}/final_{file_format}_{typology}.{file_format}', 'w')\n",
    "    for file in compact_files:\n",
    "        f.write(file)\n",
    "    f.close()\n",
    "    \n",
    "    # Eliminazione dei \n",
    "    starting_path = os.getcwd()\n",
    "    path = os.path.join(os.getcwd(), f'result_extraction/{experiment}/SRA_{typology}')\n",
    "    os.chdir(path)\n",
    "    command = 'rm *[0-9].fasta'\n",
    "    os.system(command) \n",
    "    os.chdir(starting_path)\n",
    "        \n",
    "def get_top_sequences(typology, experiment):\n",
    "    records = list(SeqIO.parse(f\"./result_extraction/{experiment}/SRA_{typology}/final_fasta_{typology}.fasta\", format=\"fasta\"))\n",
    "    print(f'Number of sequences for {typology}: {len(records)}')\n",
    "    logging.info(f'Number of sequences for {typology}: {len(records)}')\n",
    "    \n",
    "    sequences = dict()\n",
    "    for record in tqdm(records, desc='Compacting fasta'):\n",
    "        if record.seq in sequences:\n",
    "            sequences[record.seq][0] += 1\n",
    "        else:\n",
    "            sequences[record.seq] = [1, f'>{record.description}']\n",
    "    \n",
    "    print(f'Number of grouped sequences: {len(sequences)}')\n",
    "    logging.info(f'Number of grouped sequences: {len(sequences)}')\n",
    "    \n",
    "    sequences_ord = {k: v for k, v in sorted(sequences.items(), key=lambda item: item[1], reverse=True)}\n",
    "    \n",
    "    cont = 0\n",
    "    f = open(f'./result_extraction/{experiment}/SRA_{typology}/top_sequences_{typology}.fasta', 'w')\n",
    "    \n",
    "    for element in sequences_ord:\n",
    "        if sequences_ord[element][0] >= 100:\n",
    "            f.write(f'{sequences_ord[element][1]} number of reps {sequences_ord[element][0]}')\n",
    "            f.write('\\n')\n",
    "            f.write(str(element))\n",
    "            f.write('\\n')\n",
    "            cont += 1\n",
    "    f.close()\n",
    "    \n",
    "    print(f'Number of taken sequences: {cont}')\n",
    "    logging.info(f'Number of taken sequences: {cont}')\n",
    "    print()\n",
    "    logging.info('\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Main"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Carica il dataset gut\n",
    "df = pd.read_csv(\"./data/american_gut.txt\", delimiter=\"\\t\", dtype=str)\n",
    "\n",
    "# Sostituisce con NAN valori non validi\n",
    "df.replace(' ', np.nan, inplace=True)\n",
    "df.replace('Not provided', np.nan, inplace=True)\n",
    "df.replace('Unspecified', np.nan, inplace=True)\n",
    "\n",
    "# Elimina dalla working directory tutti i risultati dello scorso esperimento, ripulisce il logfile\n",
    "clean_workspace()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Inizializza il log file\n",
    "logging.basicConfig(filename='./result_extraction/sra_querying.log', level=logging.INFO, format='%(message)s')\n",
    "today = datetime.now().strftime(\"%d/%m/%Y %H:%M:%S\")\n",
    "logging.info(f'RUN TIME: {today}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## All columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for column in df.columns:\n",
    "    print(column, end = ', ')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Healthy vs not healthy study\n",
    "### Healthy extraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Estraiamo dal dataset i dati di interesse\n",
    "healthy = df.query(\"smoking_frequency == 'Never' and alcohol_frequency == 'Never'\")\n",
    "\n",
    "healthy = healthy.query(\"cancer == 'I do not have this condition'\")\n",
    "\n",
    "healthy['bmi'] = healthy['bmi'].apply(lambda x: float(x))\n",
    "healthy = healthy.query(\"bmi >= 18.5 and bmi <= 24.99\")\n",
    "\n",
    "healthy['age_years'] = healthy['age_years'].apply(lambda x: float(x))\n",
    "healthy = healthy.query(\"age_years >= 20 and age_years <= 50\")\n",
    "\n",
    "healthy = healthy.query(\"body_site == 'UBERON:feces'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sampling_data(healthy, 'healthy', 30, 'healthy_vs_not_healthy')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Not healthy extraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "not_healthy = df.query(\"smoking_frequency == 'Occasionally (1-2 times/week)' or smoking_frequency == 'Daily' or smoking_frequency == 'Regularly (3-5 times/week)'\")\n",
    "not_healthy = not_healthy.query(\"alcohol_frequency == 'Occasionally (1-2 times/week)' or alcohol_frequency == 'Daily' or alcohol_frequency == 'Regularly (3-5 times/week)'\")\n",
    "\n",
    "not_healthy = not_healthy.query(\"cancer == 'I do not have this condition'\")\n",
    "\n",
    "not_healthy['bmi'] = not_healthy['bmi'].apply(lambda x: float(x))\n",
    "not_healthy = not_healthy.query(\"bmi < 18.5 or bmi > 24.99\")\n",
    "\n",
    "not_healthy['age_years'] = not_healthy['age_years'].apply(lambda x: float(x))\n",
    "not_healthy = not_healthy.query(\"age_years >= 20 and age_years <= 50\")\n",
    "\n",
    "not_healthy = not_healthy.query(\"body_site == 'UBERON:feces'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sampling_data(not_healthy, 'not_healthy', 30, 'healthy_vs_not_healthy')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mental illness vs food disorders\n",
    "### Mental illness"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mental_illness = df.query(\"country_residence == 'United States'\")\n",
    "\n",
    "mental_illness = mental_illness.query(\"body_site == 'UBERON:feces'\")\n",
    "\n",
    "mental_illness = mental_illness.query(\"mental_illness == 'true' or mental_illness == 'Yes'\")\n",
    "\n",
    "mental_illness['age_years'] = mental_illness['age_years'].apply(lambda x: float(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sampling_data(mental_illness, 'mental_illness', 30, 'mental_ill_vs_food_dis')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Food disorders"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "food_disorders = df.query(\"country_residence == 'United States'\")\n",
    "\n",
    "food_disorders = food_disorders.query(\"body_site == 'UBERON:feces'\")\n",
    "\n",
    "food_disorders = food_disorders.query(\"mental_illness == 'false' or mental_illness == 'No'\")\n",
    "\n",
    "food_disorders['bmi'] =  food_disorders['bmi'].apply(lambda x : float(x))\n",
    "food_disorders = food_disorders.query(\"bmi < 18.5 or bmi > 24.99\")\n",
    "\n",
    "food_disorders = food_disorders.query(\"(fruit_frequency == 'Never' or fruit_frequency == 'Rarely (less than once/week)')\")\n",
    "\n",
    "food_disorders = food_disorders.query(\"exercise_frequency=='Rarely (a few times/month)' or exercise_frequency=='Never'\")\n",
    "\n",
    "food_disorders['age_years'] = food_disorders['age_years'].apply(lambda x: float(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sampling_data(food_disorders, 'food_disorders', 30, 'mental_ill_vs_food_dis')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## NCBI Quering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File number: 1\n",
      "Sample id: 10317.000067678\n",
      "Biosample ID 7497069\n",
      "Typology: mental_illness\n",
      "SRA ID: ERS1865634\n",
      "['4376103']\n",
      "\n",
      "File number: 2\n",
      "Sample id: 10317.000089968\n",
      "Biosample ID 8945722\n",
      "Typology: mental_illness\n",
      "SRA ID: ERS2409408\n",
      "['5424517']\n",
      "\n",
      "File number: 3\n",
      "Sample id: 10317.000102683\n",
      "Biosample ID 14619503\n",
      "Typology: mental_illness\n",
      "SRA ID: ERS4424358\n",
      "['10592811']\n",
      "\n",
      "File number: 4\n",
      "Sample id: 10317.000084671\n",
      "Biosample ID 8576788\n",
      "Typology: mental_illness\n",
      "SRA ID: ERS2212798\n",
      "['5144054']\n",
      "\n",
      "File number: 5\n",
      "Sample id: 10317.000074845\n",
      "Biosample ID 9541547\n",
      "Typology: mental_illness\n",
      "SRA ID: ERS2483510\n",
      "['5840876']\n",
      "\n",
      "File number: 6\n",
      "Sample id: 10317.000072390\n",
      "Biosample ID 8728606\n",
      "Typology: mental_illness\n",
      "SRA ID: ERS2303239\n",
      "['5245093']\n",
      "\n",
      "File number: 7\n",
      "Sample id: 10317.000022557\n",
      "Biosample ID 9541095\n",
      "Typology: mental_illness\n",
      "SRA ID: ERS2483066\n",
      "['5840372']\n",
      "\n",
      "File number: 8\n",
      "Sample id: 10317.000049819\n",
      "Biosample ID 5265530\n",
      "Typology: mental_illness\n",
      "SRA ID: ERS1208079\n",
      "['2650193']\n",
      "\n",
      "File number: 9\n",
      "Sample id: 10317.000036888\n",
      "Biosample ID 6367517\n",
      "Typology: mental_illness\n",
      "SRA ID: ERS1564584\n",
      "['3738053']\n",
      "\n",
      "File number: 10\n",
      "Sample id: 10317.000040397\n",
      "Biosample ID 4565258\n",
      "Typology: mental_illness\n",
      "SRA ID: ERS1088402\n",
      "['2486677', '2348292']\n",
      "\n",
      "File number: 11\n",
      "Sample id: 10317.000039595\n",
      "Biosample ID 7353796\n",
      "Typology: mental_illness\n",
      "SRA ID: ERS1821704\n",
      "['4279872']\n",
      "\n",
      "File number: 12\n",
      "Sample id: 10317.000107973\n",
      "Biosample ID 14619123\n",
      "Typology: mental_illness\n",
      "SRA ID: ERS4423978\n",
      "['10592418']\n",
      "\n",
      "File number: 13\n",
      "Sample id: 10317.000093103\n",
      "Biosample ID 9653610\n",
      "Typology: mental_illness\n",
      "SRA ID: ERS2607006\n",
      "['5960211']\n",
      "\n",
      "File number: 14\n",
      "Sample id: 10317.000067634\n",
      "Biosample ID 7497043\n",
      "Typology: mental_illness\n",
      "SRA ID: ERS1865608\n",
      "['4376077']\n",
      "\n",
      "File number: 15\n",
      "Sample id: 10317.000052139\n",
      "Biosample ID 9653056\n",
      "Typology: mental_illness\n",
      "SRA ID: ERS2606439\n",
      "['5959653']\n",
      "\n",
      "File number: 16\n",
      "Sample id: 10317.000068197\n",
      "Biosample ID 8614779\n",
      "Typology: mental_illness\n",
      "SRA ID: ERS2215143\n",
      "['5162926']\n",
      "\n",
      "File number: 17\n",
      "Sample id: 10317.000052259\n",
      "Biosample ID 6365149\n",
      "Typology: mental_illness\n",
      "SRA ID: ERS1561406\n",
      "['3735334', '3723412']\n",
      "\n",
      "File number: 18\n",
      "Sample id: 10317.000039622\n",
      "Biosample ID 7353803\n",
      "Typology: mental_illness\n",
      "SRA ID: ERS1821711\n",
      "['4279879']\n",
      "\n",
      "File number: 19\n",
      "Sample id: 10317.000079421\n",
      "Biosample ID 8577147\n",
      "Typology: mental_illness\n",
      "SRA ID: ERS2213273\n",
      "['5144414']\n",
      "\n",
      "File number: 20\n",
      "Sample id: 10317.000062984\n",
      "Biosample ID 7496826\n",
      "Typology: mental_illness\n",
      "SRA ID: ERS1865341\n",
      "['4375842']\n",
      "\n",
      "File number: 21\n",
      "Sample id: 10317.000107280\n",
      "Biosample ID 14618798\n",
      "Typology: mental_illness\n",
      "SRA ID: ERS4423956\n",
      "['10592396']\n",
      "\n",
      "File number: 22\n",
      "Sample id: 10317.000031563\n",
      "Biosample ID 7353749\n",
      "Typology: mental_illness\n",
      "SRA ID: ERS1821657\n",
      "['4279822']\n",
      "\n",
      "File number: 23\n",
      "Sample id: 10317.000089999\n",
      "Biosample ID 8945736\n",
      "Typology: mental_illness\n",
      "SRA ID: ERS2409422\n",
      "['5424531']\n",
      "\n",
      "File number: 24\n",
      "Sample id: 10317.000040070\n",
      "Biosample ID 8577218\n",
      "Typology: mental_illness\n",
      "SRA ID: ERS2213561\n",
      "['5144911']\n",
      "\n",
      "File number: 25\n",
      "Sample id: 10317.000071616\n",
      "Biosample ID 8576453\n",
      "Typology: mental_illness\n",
      "SRA ID: ERS2212475\n",
      "['5143731']\n",
      "\n",
      "File number: 26\n",
      "Sample id: 10317.000087179\n",
      "Biosample ID 8945691\n",
      "Typology: mental_illness\n",
      "SRA ID: ERS2409377\n",
      "['5424485']\n",
      "\n",
      "File number: 27\n",
      "Sample id: 10317.000069719\n",
      "Biosample ID 7439407\n",
      "Typology: mental_illness\n",
      "SRA ID: ERS1846664\n",
      "['4337518']\n",
      "\n",
      "File number: 28\n",
      "Sample id: 10317.000040411\n",
      "Biosample ID 4565263\n",
      "Typology: mental_illness\n",
      "SRA ID: ERS1088407\n",
      "['2486685', '2348297']\n",
      "\n",
      "File number: 29\n",
      "Sample id: 10317.000002929\n",
      "Biosample ID 6365469\n",
      "Typology: mental_illness\n",
      "SRA ID: ERS1561726\n",
      "['3735797']\n",
      "\n",
      "File number: 30\n",
      "Sample id: 10317.000040057\n",
      "Biosample ID 4916284\n",
      "Typology: mental_illness\n",
      "SRA ID: ERS1134839\n",
      "['2486647']\n",
      "\n",
      "File number: 31\n",
      "Sample id: 10317.000101190\n",
      "Biosample ID 14621669\n",
      "Typology: food_disorders\n",
      "SRA ID: ERS4425439\n",
      "['10594117']\n",
      "\n",
      "File number: 32\n",
      "Sample id: 10317.000067892\n",
      "Biosample ID 8614740\n",
      "Typology: food_disorders\n",
      "SRA ID: ERS2215104\n",
      "['5162885']\n",
      "\n",
      "File number: 33\n",
      "Sample id: 10317.000072842\n",
      "Biosample ID 8577547\n",
      "Typology: food_disorders\n",
      "SRA ID: ERS2213890\n",
      "['5145281']\n",
      "\n",
      "File number: 34\n",
      "Sample id: 10317.000100334\n",
      "Biosample ID 9653373\n",
      "Typology: food_disorders\n",
      "SRA ID: ERS2606755\n",
      "['5959973']\n",
      "\n",
      "File number: 35\n",
      "Sample id: 10317.000047573\n",
      "Biosample ID 6445692\n",
      "Typology: food_disorders\n",
      "SRA ID: ERS1566759\n",
      "['5840414', '3747649']\n",
      "\n",
      "File number: 36\n",
      "Sample id: 10317.000054345\n",
      "Biosample ID 6367678\n",
      "Typology: food_disorders\n",
      "SRA ID: ERS1564745\n",
      "['3738223']\n",
      "\n",
      "File number: 37\n",
      "Sample id: 10317.000093606\n",
      "Biosample ID 9653266\n",
      "Typology: food_disorders\n",
      "SRA ID: ERS2606649\n",
      "['5959867']\n",
      "\n",
      "File number: 38\n",
      "Sample id: 10317.000067987\n",
      "Biosample ID 7497205\n",
      "Typology: food_disorders\n",
      "SRA ID: ERS1865770\n",
      "['4376239']\n",
      "\n",
      "File number: 39\n",
      "Sample id: 10317.000039516\n",
      "Biosample ID 6367535\n",
      "Typology: food_disorders\n",
      "SRA ID: ERS1564602\n",
      "['3738073']\n",
      "\n",
      "File number: 40\n",
      "Sample id: 10317.000044319\n",
      "Biosample ID 5265422\n",
      "Typology: food_disorders\n",
      "SRA ID: ERS1207971\n",
      "['4933281', '2650085']\n",
      "\n",
      "File number: 41\n",
      "Sample id: 10317.000068279\n",
      "Biosample ID 7354169\n",
      "Typology: food_disorders\n",
      "SRA ID: ERS1822077\n",
      "['4280250']\n",
      "\n",
      "File number: 42\n",
      "Sample id: 10317.000094641\n",
      "Biosample ID 9653665\n",
      "Typology: food_disorders\n",
      "SRA ID: ERS2607061\n",
      "['5960267']\n",
      "\n",
      "File number: 43\n",
      "Sample id: 10317.000090080\n",
      "Biosample ID 8945766\n",
      "Typology: food_disorders\n",
      "SRA ID: ERS2409452\n",
      "['5424561']\n",
      "\n",
      "File number: 44\n",
      "Sample id: 10317.000079461\n",
      "Biosample ID 8569597\n",
      "Typology: food_disorders\n",
      "SRA ID: ERS2210311\n",
      "['5134484']\n",
      "\n",
      "File number: 45\n",
      "Sample id: 10317.000065701\n",
      "Biosample ID 7354080\n",
      "Typology: food_disorders\n",
      "SRA ID: ERS1821988\n",
      "['4280161']\n",
      "\n",
      "File number: 46\n",
      "Sample id: 10317.000068295\n",
      "Biosample ID 7496890\n",
      "Typology: food_disorders\n",
      "SRA ID: ERS1865405\n",
      "['4375906']\n",
      "\n",
      "File number: 47\n",
      "Sample id: 10317.000039742\n",
      "Biosample ID 8614211\n",
      "Typology: food_disorders\n",
      "SRA ID: ERS2214221\n",
      "['5162315']\n",
      "\n",
      "File number: 48\n",
      "Sample id: 10317.000072547\n",
      "Biosample ID 14618547\n",
      "Typology: food_disorders\n",
      "SRA ID: ERS4423705\n",
      "['10592086']\n",
      "\n",
      "File number: 49\n",
      "Sample id: 10317.000074360\n",
      "Biosample ID 8728634\n",
      "Typology: food_disorders\n",
      "SRA ID: ERS2303267\n",
      "['5245121']\n",
      "\n",
      "File number: 50\n",
      "Sample id: 10317.000082850\n",
      "Biosample ID 8729102\n",
      "Typology: food_disorders\n",
      "SRA ID: ERS2303401\n",
      "['5245256']\n",
      "\n",
      "File number: 51\n",
      "Sample id: 10317.000052019\n",
      "Biosample ID 8945351\n",
      "Typology: food_disorders\n",
      "SRA ID: ERS2409145\n",
      "['5424253']\n",
      "\n",
      "File number: 52\n",
      "Sample id: 10317.000059163\n",
      "Biosample ID 6365184\n",
      "Typology: food_disorders\n",
      "SRA ID: ERS1561441\n",
      "['3735370', '3723480']\n",
      "\n",
      "File number: 53\n",
      "Sample id: 10317.000076751\n",
      "Biosample ID 8569121\n",
      "Typology: food_disorders\n",
      "SRA ID: ERS2203295\n",
      "['5125468']\n",
      "\n",
      "File number: 54\n",
      "Sample id: 10317.000071590\n",
      "Biosample ID 8614879\n",
      "Typology: food_disorders\n",
      "SRA ID: ERS2215243\n",
      "['5163032']\n",
      "\n",
      "File number: 55\n",
      "Sample id: 10317.000087191\n",
      "Biosample ID 9653540\n",
      "Typology: food_disorders\n",
      "SRA ID: ERS2606936\n",
      "['5960140']\n",
      "\n",
      "File number: 56\n",
      "Sample id: 10317.000073874\n",
      "Biosample ID 8614492\n",
      "Typology: food_disorders\n",
      "SRA ID: ERS2214502\n",
      "['5162601']\n",
      "\n",
      "File number: 57\n",
      "Sample id: 10317.000084291\n",
      "Biosample ID 8576667\n",
      "Typology: food_disorders\n",
      "SRA ID: ERS2212677\n",
      "['5143933']\n",
      "\n",
      "File number: 58\n",
      "Sample id: 10317.000072255\n",
      "Biosample ID 9541379\n",
      "Typology: food_disorders\n",
      "SRA ID: ERS2483345\n",
      "['5840707']\n",
      "\n",
      "File number: 59\n",
      "Sample id: 10317.000062971\n",
      "Biosample ID 7496825\n",
      "Typology: food_disorders\n",
      "SRA ID: ERS1865340\n",
      "['4375841']\n",
      "\n",
      "File number: 60\n",
      "Sample id: 10317.000050338\n",
      "Biosample ID 14618671\n",
      "Typology: food_disorders\n",
      "SRA ID: ERS4423829\n",
      "['10592264']\n",
      "\n",
      "Number of sequences for mental_illness: 903794\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "7b6adcf9e40e4093ab02ff9783704217",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, description='Compacting fasta', max=903794.0, style=ProgressStyle(desc…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Number of grouped sequences: 211619\n",
      "Number of taken sequences: 550\n",
      "\n",
      "Number of sequences for food_disorders: 1549729\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "31b49062dec740d79580c82c7e18c297",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, description='Compacting fasta', max=1549729.0, style=ProgressStyle(des…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Number of grouped sequences: 867911\n",
      "Number of taken sequences: 564\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sra_querying('mental_ill_vs_food_dis', ['mental_illness', 'food_disorders'])\n",
    "\n",
    "# 7497069, 14618671"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File number: 1\n",
      "Sample id: 10317.000093022\n",
      "Biosample ID 9653255\n",
      "Typology: healthy\n",
      "SRA ID: ERS2606638\n",
      "['5959856']\n",
      "\n",
      "File number: 2\n",
      "Sample id: 10317.000058340\n",
      "Biosample ID 6367687\n",
      "Typology: healthy\n",
      "SRA ID: ERS1564754\n",
      "['3738232']\n",
      "\n",
      "File number: 3\n",
      "Sample id: 10317.000107254\n",
      "Biosample ID 14618791\n",
      "Typology: healthy\n",
      "SRA ID: ERS4423949\n",
      "['10592389']\n",
      "\n",
      "File number: 4\n",
      "Sample id: 10317.000032796\n",
      "Biosample ID 5158596\n",
      "Typology: healthy\n",
      "SRA ID: ERS1164403\n",
      "['4933109', '2559054']\n",
      "\n",
      "File number: 5\n",
      "Sample id: 10317.000046476\n",
      "Biosample ID 7353916\n",
      "Typology: healthy\n",
      "SRA ID: ERS1821824\n",
      "['4279996']\n",
      "\n",
      "File number: 6\n",
      "Sample id: 10317.000079781\n",
      "Biosample ID 8569142\n",
      "Typology: healthy\n",
      "SRA ID: ERS2203316\n",
      "['5125489']\n",
      "\n",
      "File number: 7\n",
      "Sample id: 10317.000040396\n",
      "Biosample ID 6364886\n",
      "Typology: healthy\n",
      "SRA ID: ERS1561319\n",
      "['3735156', '3723191']\n",
      "\n",
      "File number: 8\n",
      "Sample id: 10317.000069615\n",
      "Biosample ID 8614819\n",
      "Typology: healthy\n",
      "SRA ID: ERS2215183\n",
      "['5162969']\n",
      "\n",
      "File number: 9\n",
      "Sample id: 10317.000050243\n",
      "Biosample ID 6365125\n",
      "Typology: healthy\n",
      "SRA ID: ERS1561382\n",
      "['3735298', '3723373']\n",
      "\n",
      "File number: 10\n",
      "Sample id: 10317.000092617\n",
      "Biosample ID 9653598\n",
      "Typology: healthy\n",
      "SRA ID: ERS2606994\n",
      "['5960199']\n",
      "\n",
      "File number: 11\n",
      "Sample id: 10317.000043083\n",
      "Biosample ID 4916368\n",
      "Typology: healthy\n",
      "SRA ID: ERS1134923\n",
      "['2486755']\n",
      "\n",
      "File number: 12\n",
      "Sample id: 10317.000058882\n",
      "Biosample ID 8614242\n",
      "Typology: healthy\n",
      "SRA ID: ERS2214252\n",
      "['5162347']\n",
      "\n",
      "File number: 13\n",
      "Sample id: 10317.000039907\n",
      "Biosample ID 4565184\n",
      "Typology: healthy\n",
      "SRA ID: ERS1088328\n",
      "['2348218']\n",
      "\n",
      "File number: 14\n",
      "Sample id: 10317.000029546\n",
      "Biosample ID 5158551\n",
      "Typology: healthy\n",
      "SRA ID: ERS1164358\n",
      "['2559008']\n",
      "\n",
      "File number: 15\n",
      "Sample id: 10317.000068197\n",
      "Biosample ID 8614779\n",
      "Typology: healthy\n",
      "SRA ID: ERS2215143\n",
      "['5162926']\n",
      "\n",
      "File number: 16\n",
      "Sample id: 10317.000072168\n",
      "Biosample ID 9541357\n",
      "Typology: healthy\n",
      "SRA ID: ERS2483323\n",
      "['5840684']\n",
      "\n",
      "File number: 17\n",
      "Sample id: 10317.000072038\n",
      "Biosample ID 8945382\n",
      "Typology: healthy\n",
      "SRA ID: ERS2409176\n",
      "['5424284']\n",
      "\n",
      "File number: 18\n",
      "Sample id: 10317.000065690\n",
      "Biosample ID 7354073\n",
      "Typology: healthy\n",
      "SRA ID: ERS1821981\n",
      "['4280154']\n",
      "\n",
      "File number: 19\n",
      "Sample id: 10317.000067863\n",
      "Biosample ID 7353609\n",
      "Typology: healthy\n",
      "SRA ID: ERS1821515\n",
      "['4279681']\n",
      "\n",
      "File number: 20\n",
      "Sample id: 10317.000046475\n",
      "Biosample ID 7353915\n",
      "Typology: healthy\n",
      "SRA ID: ERS1821823\n",
      "['4279995']\n",
      "\n",
      "File number: 21\n",
      "Sample id: 10317.000058379\n",
      "Biosample ID 6367696\n",
      "Typology: healthy\n",
      "SRA ID: ERS1564763\n",
      "['3738241']\n",
      "\n",
      "File number: 22\n",
      "Sample id: 10317.000068199\n",
      "Biosample ID 7354474\n",
      "Typology: healthy\n",
      "SRA ID: ERS1822469\n",
      "['4280575']\n",
      "\n",
      "File number: 23\n",
      "Sample id: 10317.000087274\n",
      "Biosample ID 9540903\n",
      "Typology: healthy\n",
      "SRA ID: ERS2482874\n",
      "['5840138']\n",
      "\n",
      "File number: 24\n",
      "Sample id: 10317.000074433\n",
      "Biosample ID 9541535\n",
      "Typology: healthy\n",
      "SRA ID: ERS2483498\n",
      "['5840864']\n",
      "\n",
      "File number: 25\n",
      "Sample id: 10317.000041681\n",
      "Biosample ID 5265305\n",
      "Typology: healthy\n",
      "SRA ID: ERS1207854\n",
      "['2649968']\n",
      "\n",
      "File number: 26\n",
      "Sample id: 10317.000097633\n",
      "Biosample ID 9657761\n",
      "Typology: healthy\n",
      "SRA ID: ERS2608455\n",
      "['5964456']\n",
      "\n",
      "File number: 27\n",
      "Sample id: 10317.000100377\n",
      "Biosample ID 9653382\n",
      "Typology: healthy\n",
      "SRA ID: ERS2606764\n",
      "['5959982']\n",
      "\n",
      "File number: 28\n",
      "Sample id: 10317.000097102\n",
      "Biosample ID 9653687\n",
      "Typology: healthy\n",
      "SRA ID: ERS2607083\n",
      "['5960289']\n",
      "\n",
      "File number: 29\n",
      "Sample id: 10317.000065672\n",
      "Biosample ID 7354060\n",
      "Typology: healthy\n",
      "SRA ID: ERS1821968\n",
      "['4280141']\n",
      "\n",
      "File number: 30\n",
      "Sample id: 10317.000058955\n",
      "Biosample ID 6445726\n",
      "Typology: healthy\n",
      "SRA ID: ERS1566793\n",
      "['3747685']\n",
      "\n",
      "File number: 31\n",
      "Sample id: 10317.000054208\n",
      "Biosample ID 6365889\n",
      "Typology: not_healthy\n",
      "SRA ID: ERS1562146\n",
      "['3736270']\n",
      "\n",
      "File number: 32\n",
      "Sample id: 10317.000070659\n",
      "Biosample ID 8614384\n",
      "Typology: not_healthy\n",
      "SRA ID: ERS2214394\n",
      "['5162493']\n",
      "\n",
      "File number: 33\n",
      "Sample id: 10317.000050274\n",
      "Biosample ID 6366103\n",
      "Typology: not_healthy\n",
      "SRA ID: ERS1562360\n",
      "['3736485']\n",
      "\n",
      "File number: 34\n",
      "Sample id: 10317.000042630\n",
      "Biosample ID 5709882\n",
      "Typology: not_healthy\n",
      "SRA ID: ERS1305444\n",
      "['3047637']\n",
      "\n",
      "File number: 35\n",
      "Sample id: 10317.000093103\n",
      "Biosample ID 9653610\n",
      "Typology: not_healthy\n",
      "SRA ID: ERS2607006\n",
      "['5960211']\n",
      "\n",
      "File number: 36\n",
      "Sample id: 10317.000059040\n",
      "Biosample ID 6366200\n",
      "Typology: not_healthy\n",
      "SRA ID: ERS1562457\n",
      "['3736582']\n",
      "\n",
      "File number: 37\n",
      "Sample id: 10317.000068175\n",
      "Biosample ID 7354470\n",
      "Typology: not_healthy\n",
      "SRA ID: ERS1822465\n",
      "['4280571']\n",
      "\n",
      "File number: 38\n",
      "Sample id: 10317.000047222\n",
      "Biosample ID 6366062\n",
      "Typology: not_healthy\n",
      "SRA ID: ERS1562319\n",
      "['3736444']\n",
      "\n",
      "File number: 39\n",
      "Sample id: 10317.000093400\n",
      "Biosample ID 9541016\n",
      "Typology: not_healthy\n",
      "SRA ID: ERS2482987\n",
      "['5840252']\n",
      "\n",
      "File number: 40\n",
      "Sample id: 10317.000075909\n",
      "Biosample ID 8945523\n",
      "Typology: not_healthy\n",
      "SRA ID: ERS2409209\n",
      "['5424317']\n",
      "\n",
      "File number: 41\n",
      "Sample id: 10317.000037933\n",
      "Biosample ID 6365537\n",
      "Typology: not_healthy\n",
      "SRA ID: ERS1561794\n",
      "['3735881', '4337162']\n",
      "\n",
      "File number: 42\n",
      "Sample id: 10317.000038160\n",
      "Biosample ID 8577206\n",
      "Typology: not_healthy\n",
      "SRA ID: ERS2213549\n",
      "['5144897']\n",
      "\n",
      "File number: 43\n",
      "Sample id: 10317.000058971\n",
      "Biosample ID 9653062\n",
      "Typology: not_healthy\n",
      "SRA ID: ERS2606445\n",
      "['5959659']\n",
      "\n",
      "File number: 44\n",
      "Sample id: 10317.000105372\n",
      "Biosample ID 14618764\n",
      "Typology: not_healthy\n",
      "SRA ID: ERS4423922\n",
      "['10592362']\n",
      "\n",
      "File number: 45\n",
      "Sample id: 10317.000047151\n",
      "Biosample ID 6365612\n",
      "Typology: not_healthy\n",
      "SRA ID: ERS1561869\n",
      "['3735977']\n",
      "\n",
      "File number: 46\n",
      "Sample id: 10317.000070708\n",
      "Biosample ID 8728589\n",
      "Typology: not_healthy\n",
      "SRA ID: ERS2303222\n",
      "['5245076']\n",
      "\n",
      "File number: 47\n",
      "Sample id: 10317.000107269\n",
      "Biosample ID 14618796\n",
      "Typology: not_healthy\n",
      "SRA ID: ERS4423954\n",
      "['10592394']\n",
      "\n",
      "File number: 48\n",
      "Sample id: 10317.000068174\n",
      "Biosample ID 7354469\n",
      "Typology: not_healthy\n",
      "SRA ID: ERS1822464\n",
      "['4280570']\n",
      "\n",
      "File number: 49\n",
      "Sample id: 10317.000041832\n",
      "Biosample ID 6366029\n",
      "Typology: not_healthy\n",
      "SRA ID: ERS1562286\n",
      "['3736411']\n",
      "\n",
      "File number: 50\n",
      "Sample id: 10317.000065320\n",
      "Biosample ID 6445740\n",
      "Typology: not_healthy\n",
      "SRA ID: ERS1566807\n",
      "['3747699']\n",
      "\n",
      "File number: 51\n",
      "Sample id: 10317.000047463\n",
      "Biosample ID 7353923\n",
      "Typology: not_healthy\n",
      "SRA ID: ERS1821831\n",
      "['4280003']\n",
      "\n",
      "File number: 52\n",
      "Sample id: 10317.000093404\n",
      "Biosample ID 9541017\n",
      "Typology: not_healthy\n",
      "SRA ID: ERS2482988\n",
      "['5840254']\n",
      "\n",
      "File number: 53\n",
      "Sample id: 10317.000060228\n",
      "Biosample ID 8614699\n",
      "Typology: not_healthy\n",
      "SRA ID: ERS2215063\n",
      "['5162829']\n",
      "\n",
      "File number: 54\n",
      "Sample id: 10317.000069197\n",
      "Biosample ID 14618489\n",
      "Typology: not_healthy\n",
      "SRA ID: ERS4423647\n",
      "['10592027']\n",
      "\n",
      "File number: 55\n",
      "Sample id: 10317.000068657\n",
      "Biosample ID 8569476\n",
      "Typology: not_healthy\n",
      "SRA ID: ERS2210190\n",
      "['5134363']\n",
      "\n",
      "File number: 56\n",
      "Sample id: 10317.000105371\n",
      "Biosample ID 14618763\n",
      "Typology: not_healthy\n",
      "SRA ID: ERS4423921\n",
      "['10592361']\n",
      "\n",
      "File number: 57\n",
      "Sample id: 10317.000075933\n",
      "Biosample ID 9653092\n",
      "Typology: not_healthy\n",
      "SRA ID: ERS2606475\n",
      "['5959691']\n",
      "\n",
      "File number: 58\n",
      "Sample id: 10317.000062086\n",
      "Biosample ID 6365939\n",
      "Typology: not_healthy\n",
      "SRA ID: ERS1562196\n",
      "['3736320']\n",
      "\n",
      "File number: 59\n",
      "Sample id: 10317.000101067\n",
      "Biosample ID 14621574\n",
      "Typology: not_healthy\n",
      "SRA ID: ERS4425344\n",
      "['10594022']\n",
      "\n",
      "File number: 60\n",
      "Sample id: 10317.000038305\n",
      "Biosample ID 4916250\n",
      "Typology: not_healthy\n",
      "SRA ID: ERS1134805\n",
      "['2486602']\n",
      "\n",
      "Number of sequences for healthy: 1833387\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "7f883b2fe8234e0eb09b728049c49b51",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, description='Compacting fasta', max=1833387.0, style=ProgressStyle(des…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Number of grouped sequences: 1184492\n",
      "Number of taken sequences: 547\n",
      "\n",
      "Number of sequences for not_healthy: 858682\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "5441ae39225046ecad767b598eec5221",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, description='Compacting fasta', max=858682.0, style=ProgressStyle(desc…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Number of grouped sequences: 260130\n",
      "Number of taken sequences: 464\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sra_querying('healthy_vs_not_healthy', ['healthy', 'not_healthy'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "final_project_bio",
   "language": "python",
   "name": "final_project_bio"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
