{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIWWW\n",
    "from Bio import SeqIO\n",
    "import os\n",
    "import Bio.KEGG.REST as REST\n",
    "import pandas as pd\n",
    "\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import Entrez\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getOrgs():\n",
    "    Org = REST.kegg_list('organism')\n",
    "    dfAllOrgs = pd.DataFrame(columns=['Tnumber', 'orgCode', 'orgName', 'phylogeny'])\n",
    "    for r in Org.readlines():\n",
    "        spltOrg = r.split(\"\\t\")\n",
    "        sngItem = pd.DataFrame([spltOrg], columns=['Tnumber', 'orgCode', 'orgName', 'phylogeny'])\n",
    "        dfAllOrgs = dfAllOrgs.append(sngItem, ignore_index=True)\n",
    "    return dfAllOrgs\n",
    "\n",
    "##### FROM BYPYTHON, IT IS NOT MINE\n",
    "from Bio._py3k import urlopen as _urlopen\n",
    "from Bio._py3k import _binary_to_string_handle\n",
    "\n",
    "\n",
    "def _q(op, arg1, arg2=None, arg3=None):\n",
    "    URL = \"http://rest.kegg.jp/%s\"\n",
    "    if arg2 and arg3:\n",
    "        args = \"%s/%s/%s/%s\" % (op, arg1, arg2, arg3)\n",
    "    elif arg2:\n",
    "        args = \"%s/%s/%s\" % (op, arg1, arg2)\n",
    "    else:\n",
    "        args = \"%s/%s\" % (op, arg1)\n",
    "    resp = _urlopen(URL % (args))\n",
    "\n",
    "    if \"image\" == arg2:\n",
    "        return resp\n",
    "\n",
    "    return _binary_to_string_handle(resp)\n",
    "\n",
    "\n",
    "\n",
    "def kegg_info(database):\n",
    "\n",
    "    return _q(\"info\", database)\n",
    "\n",
    "\n",
    "def kegg_list(database, org=None):\n",
    "    if isinstance(database, str) and (database in [\"pathway\", \"module\"]) and org:\n",
    "        resp = _q(\"list\", database, org)\n",
    "    elif isinstance(database, str) and database and org:\n",
    "        raise Exception(\"Invalid database arg for kegg list request.\")\n",
    "    else:\n",
    "        if isinstance(database, list) and len(database) <= 100:\n",
    "            database = (\"+\").join(database)\n",
    "        elif isinstance(database, list) and len(database) > 100:\n",
    "            raise Exception(\"Maximuim number of databases is 100 for kegg list query\")\n",
    "        resp = _q(\"list\", database)\n",
    "\n",
    "    return resp\n",
    "\n",
    "\n",
    "def kegg_find(database, query, option=None):\n",
    "    if database in [\"compound\", \"drug\"] and \\\n",
    "          option in [\"formula\", \"exact_mass\", \"mol_weight\"]:\n",
    "        resp = _q(\"find\", database, query, option)\n",
    "    elif option:\n",
    "        raise Exception(\"Invalid option arg for kegg find request.\")\n",
    "    else:\n",
    "        if isinstance(query, list):\n",
    "            query = \"+\".join(query)\n",
    "        resp = _q(\"find\", database, query)\n",
    "\n",
    "    return resp\n",
    "\n",
    "\n",
    "def kegg_get(dbentries, option=None):\n",
    "    if isinstance(dbentries, list) and len(dbentries) <= 10:\n",
    "        dbentries = \"+\".join(dbentries)\n",
    "    elif isinstance(dbentries, list) and len(dbentries) > 10:\n",
    "        raise Exception(\"Maximum number of dbentries is 10 for kegg get query\")\n",
    "    if option in [\"aaseq\", \"ntseq\", \"mol\", \"kcf\", \"image\", \"kgml\"]:\n",
    "        resp = _q(\"get\", dbentries, option)\n",
    "    elif option:\n",
    "        raise Exception(\"Invalid option arg for kegg get request.\")\n",
    "    else:\n",
    "        resp = _q(\"get\", dbentries)\n",
    "\n",
    "    return resp\n",
    "\n",
    "\n",
    "def kegg_conv(target_db, source_db, option=None):\n",
    "    if option and option not in [\"turtle\", \"n-triple\"]:\n",
    "        raise Exception(\"Invalid option arg for kegg conv request.\")\n",
    "\n",
    "    if isinstance(source_db, list):\n",
    "        source_db = \"+\".join(source_db)\n",
    "\n",
    "    if target_db in [\"ncbi-gi\", \"ncbi-geneid\", \"ncbi-proteinid\", \"uniprot\", \"genes\"] or \\\n",
    "       source_db in [\"ncbi-gi\", \"ncbi-geneid\", \"ncbi-proteinid\", \"uniprot\", \"genes\"] or \\\n",
    "       (target_db in [\"drug\", \"compound\", \"glycan\"] and\n",
    "           source_db in [\"pubchem\", \"glycan\"]) or \\\n",
    "       (target_db in [\"pubchem\", \"glycan\"] and\n",
    "           source_db in [\"drug\", \"compound\", \"glycan\"]):\n",
    "\n",
    "        if option:\n",
    "            resp = _q(\"conv\", target_db, source_db, option)\n",
    "        else:\n",
    "            resp = _q(\"conv\", target_db, source_db)\n",
    "\n",
    "        return resp\n",
    "    else:\n",
    "        raise Exception(\"Bad argument target_db or source_db for kegg conv request.\")\n",
    "\n",
    "\n",
    "def kegg_link(target_db, source_db, option=None):\n",
    "    if option and option not in [\"turtle\", \"n-triple\"]:\n",
    "        raise Exception(\"Invalid option arg for kegg conv request.\")\n",
    "\n",
    "    if isinstance(source_db, list):\n",
    "        source_db = \"+\".join(source_db)\n",
    "\n",
    "    if option:\n",
    "        resp = _q(\"link\", target_db, source_db, option)\n",
    "    else:\n",
    "        resp = _q(\"link\", target_db, source_db)\n",
    "\n",
    "    return resp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "record = SeqIO.read(\"./data/anisakis_coi.fasta\", format=\"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Output 1:\n",
      " CATCCTGAGGTTTATATTTTGATTTTACCGGCTTTTGGTATTATTAGACAGTCTAGTTTGTATTTAACTGGTAAGAAAGAAGTTTTTGGTTCTTTGGGGATGGTATATGCTATTTTAAGAATTGGTCTTATTGGTTGTGTAGTTTGGGCTCATCATATGTATACTGTGGGCATGGATCTTGATTCTCGTGCTTATTTTACTGCTGCTACTATGGTTATTGCTGTTCCTACTGGGGTAAAAGTTTTTAGTTGATTAGCTACTCTTTTTGGTATGAAGATGGTTTTTCAACCTTTGCTTCTTTGGGTATTGGGCTTTATCTTTTTATTTACTGTTGGTGGTTTGACTGGTGTTGTTCTTTCTAATTCTAGTTTGGATGTTATTCTGCATGATACTTATTATGTAGTAAGTCATTTTCATTATGT \n",
      "\n",
      "Output 2:\n",
      " >KY973682.1 Anisakis simplex cytochrome c oxidase subunit I (COI) gene, partial cds; mitochondrial\n",
      "CATCCTGAGGTTTATATTTTGATTTTACCGGCTTTTGGTATTATTAGACAGTCTAGTTTG\n",
      "TATTTAACTGGTAAGAAAGAAGTTTTTGGTTCTTTGGGGATGGTATATGCTATTTTAAGA\n",
      "ATTGGTCTTATTGGTTGTGTAGTTTGGGCTCATCATATGTATACTGTGGGCATGGATCTT\n",
      "GATTCTCGTGCTTATTTTACTGCTGCTACTATGGTTATTGCTGTTCCTACTGGGGTAAAA\n",
      "GTTTTTAGTTGATTAGCTACTCTTTTTGGTATGAAGATGGTTTTTCAACCTTTGCTTCTT\n",
      "TGGGTATTGGGCTTTATCTTTTTATTTACTGTTGGTGGTTTGACTGGTGTTGTTCTTTCT\n",
      "AATTCTAGTTTGGATGTTATTCTGCATGATACTTATTATGTAGTAAGTCATTTTCATTAT\n",
      "GT\n",
      " \n",
      "\n",
      "Length: 422\n"
     ]
    }
   ],
   "source": [
    "print('Output 1:\\n', record.seq, '\\n')\n",
    "print('Output 2:\\n', record.format(\"fasta\"), '\\n')\n",
    "print(f'Length: {len(record.seq)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "blastResult = NCBIWWW.qblast(\"blastx\", \"nucleotide\", record.format(\"fasta\")) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f'./data/blastOutput_{str(record.id)}.xml', \"w\") as f:\n",
    "    f.write(blastResult.read())\n",
    "    blastResult.close()\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfAllOrgsInfo = getOrgs()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfAllOrgsInfo.to_excel('./data/dfOrgs.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Entrez.email = 'giacomo.villa.mi@gmail.com'\n",
    "\n",
    "dfOrgs = pd.read_excel(\"./data/dfOrgs.xlsx\")\n",
    "\n",
    "queryNames = []\n",
    "ncbiIDs = []\n",
    "descriptions = []\n",
    "orgNames = []\n",
    "expvalues = []\n",
    "lengths = []\n",
    "keggIds = []\n",
    "\n",
    "result_handle = open(f'./data/blastOutput_{str(record.id)}.xml', \"r\")\n",
    "\n",
    "blast_records = NCBIXML.read(result_handle)\n",
    "\n",
    "queryName = blast_records.query\n",
    "\n",
    "aligns = blast_records.alignments\n",
    "\n",
    "for hit in aligns:       \n",
    "    ncbiID = hit.accession #The NCBI identifier\n",
    "    description = hit.hit_def #The description of the sequence\n",
    "    length = hit.length\n",
    "    hsp = hit.hsps\n",
    "    \n",
    "    entrez_result = Entrez.efetch(db=\"protein\", id=ncbiID, rettype=\"gp\", retmode=\"xml\")\n",
    "    \n",
    "    record = Entrez.read(entrez_result), '\\n'\n",
    "    \n",
    "    # print('record\\n',record[0], '\\n\\n', record[0][0], '\\n\\n', record[0][0]['GBSeq_organism'],'\\n\\n')\n",
    "    \n",
    "    orgName = record[0][0]['GBSeq_organism']\n",
    "    print('NCBI ID: ', ncbiID)\n",
    "    print('Description: ', description)\n",
    "    print('Sequence length: ', length)\n",
    "    evalues = []\n",
    "    for h in hsp:\n",
    "        evalue = h.expect #E-value of each hsp\n",
    "        evalues.append(evalue)\n",
    "        print('Evalue: ', evalues)\n",
    "    print('Organism: ', orgName)    \n",
    "    orgCode = ''    \n",
    "    for i, row in dfOrgs.iterrows():\n",
    "        lsname = dfOrgs['orgName'][i].split(\" \")[0:2]\n",
    "        name = lsname[0] + \" \" + lsname[1]\n",
    "        if name in orgName:\n",
    "            orgCode = dfOrgs['orgCode'][i]\n",
    "    print('Organism code: ', orgCode)\n",
    "    keggId = ''\n",
    "    #OrgCodes.append(orgCode)\n",
    "    ## Conversion of RefSeq Protein ID to KEGG ID\n",
    "    if orgCode != '':\n",
    "        keggSearch = kegg_conv(orgCode, \"ncbi-proteinid\")\n",
    "        print('keggSearch\\n', keggSearch, '\\n')\n",
    "        ncbiAccessionNumber = 'ncbi-proteinid:' + ncbiID\n",
    "        for k in keggSearch:\n",
    "            print('k\\n', k, '\\n')\n",
    "            ncbi = k.split(\"\\n\")[0].split(\"\\t\")[0]\n",
    "            kegg = k.split(\"\\n\")[0].split(\"\\t\")[1]\n",
    "            if ncbi == ncbiAccessionNumber:\n",
    "                keggId = kegg\n",
    "                #keggIDs.append(kegg)\n",
    "                break\n",
    "    print('KEGG ID: ', keggId, '\\n')\n",
    "    print()\n",
    "    queryNames.append(queryName)\n",
    "    ncbiIDs.append(ncbiID)\n",
    "    descriptions.append(description)\n",
    "    orgNames.append(orgName)\n",
    "    expvalues.append(evalues)\n",
    "    lengths.append(length)\n",
    "    keggIds.append(keggId)\n",
    "\n",
    "dfinfo = pd.DataFrame({'QueryName': queryNames, 'Description': descriptions, 'E-value': expvalues,\n",
    "                      'NCBI ID': ncbiIDs, 'KEGG ID': keggIds, 'Organism': orgNames,  'Length':lengths})\n",
    "dfinfo = dfinfo[['QueryName', 'Description', 'E-value','NCBI ID', 'KEGG ID', 'Organism', 'Length']]\n",
    "dfinfo.to_excel(f'./data/blastOutput_{str(record.id)}.xml')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biosciences",
   "language": "python",
   "name": "biosciences"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
